executor/update.go

```diff
@@ -62,6 +63,11 @@ func (e *UpdateExec) exec(schema *expression.Schema) ([]types.Datum, error) {
 		for _, col := range cols {
 			offset := getTableOffset(schema, col)
 			end := offset + len(tbl.WritableCols())
+			handleDatum := row[col.Index]
+			// handleDatum is nil only when outer join fill no matched columns.
```

Comment:
How about wrapping a function to check whether we can update the current record:
```go
// canNotUpdate checks the handle of a record to decide whether that record
// can not be updated. The handle is NULL only when it is the inner side of an
// outer join: the outer row can not match any inner rows, and in this scenario
// the inner handle field is filled with a NULL value.
//
// This fixes: https://github.com/pingcap/tidb/issues/7176.
func (e *UpdateExec) canNotUpdate(handle types.Datum) bool {
    return handle.IsNull()
}
```